### PR TITLE
feat: built-in content blocking based on IPIP-383

### LINKED ISSUE #8492: IPFS filtering to allow node operators to decide on content they are willing to serve
### Checklist

- [X] My issue is specific & actionable.
- [X] I am not suggesting a protocol enhancement.
- [X] I have searched on the [issue tracker](https://github.com/ipfs/go-ipfs/issues?q=is%3Aissue) for my issue.

### Description

Recently, Cloudflare has open sourced a fork of [go-ipfs](https://github.com/cloudflare/go-ipfs/tree/v0.9.1-safemode) providing filtering capabilities, grouped under `safemode` command. The architecture is described in a [dedicated blog](https://blog.cloudflare.com/cloudflare-ipfs-safe-mode/).

The system works by filtering certain CID when walking the DAG. This allow node operators to prevent certain CID from being provided, both by the HTTP gateway and to the P2P network.
CIDs to be filtered are stored in a [blocklist](https://github.com/cloudflare/go-ipfs-blocklist). By default, this blocklist is in a dedicated mount of the datastore `/safemode`.

Action that can be performed by a blocklist are (based on the [proposed interface](https://github.com/cloudflare/go-ipfs-blocklist/blob/main/blocklist.go#L20)):
+ `block` to add content to the blocklist
+ `unblock` to remove it
+ `purge` to remove content from the blockstore. Ideally, this option could be extensible, to purge remote datastore, or HTTP cache for instance
+ `search` to query the blocklist
+ `audit` to access the log of actions that have been performed against the blocklist

For convenience, `ipfs safemode` command provides multiple way to resolve content. From its documentation:
```
- IPFS address, i.e. /ipfs/<CID>
- IPNS address, i.e. /ipns/<hash_publickey>
- DNSLink address, i.e. /ipns/example.com
- HTTP URL, i.e. https://example.com/ or https://gateway.example.com/ipfs/<CID>
```

This is a proposal implementation, which satisfies some requirements laid out in https://github.com/ipfs/roadmap/issues/64. It provides a more standardised approach for node operators to filter content they are willing to provide.

The implementation has been developed 3 years ago, and may not suit the current architecture of the go-ipfs project.

### PR DESCRIPTION
**Pull request by hsanjuan**
Fixes #8492.

This introduces "nopfs" as a preloaded plugin into Kubo.

It automatically make Kubo watch *.deny files (syntax from [IPIP-383](https://github.com/ipfs/specs/pull/383)) found in:

- `/etc/ipfs/denylists`
- `$XDG_CONFIG_HOME/ipfs/denylists`
- `$IPFS_PATH/denylists`

(files need to be present before starting in order to be watched).

Debug logging can be enabled with `GOLOG_LOG_LEVEL="nopfs=debug"`.

All blocks are logged to "nopfs-blocks", so logging requests for blocked content can be achieved with
`GOLOG_LOG_LEVEL="nopfs-blocks=warn"`:

```
WARN (...) QmRFniDxwxoG2n4AcnGhRdjqDjCM5YeUcBE75K8WXmioH3: blocked (test.deny:9)
```

Interactive/gateway users will also receive errors as responses but with less details:

```
Error: /ipfs/QmQvjk82hPkSaZsyJ8vNER5cmzKW7HyGX5XVusK7EAenCN is blocked and cannot be provided
```

One particularity to keep in mind is that GetMany() will silently drop blocked blocks from the response (warnings are logged). AddMany() will act similarly and avoid adding blocked blocks.

The code implementing all this is actually in nopfs:

- https://github.com/ipfs-shipyard/nopfs (main library)
- https://github.com/ipfs-shipyard/nopfs/tree/master/ipfs (wrappers)

The interpretation of the list rules and block detection is well tested, but a general review might be in order.

### TODO

- [x] merge  nopfs PRs
  - [x] https://github.com/ipfs-shipyard/nopfs/pull/25
  - [x] https://github.com/ipfs-shipyard/nopfs/pull/28
- [x] add basic CLI regression tests (`test/cli/content_blocking_test.go`)
- [x] add basic Gateway regression tests (`test/cli/content_blocking_test.go`)
- [x] dry tests so we define testCases and commands as lists and run matrix
  - [x] make `dag export` accept CID _or_ Path like all other commands
- [x] include special edge case tests for subdomain gateway
- [x] resolve remaining TODO from `test/cli/content_blocking_test.go`
  - [x] when blocked, HTTP error code is 500, should be 410 
    - [x] https://github.com/ipfs/boxo/pull/497
  - [x] "no fetch" gateway tests should pass on CI (https://github.com/ipfs/kubo/pull/10161#issuecomment-1782175955)

### CODE REVIEW

File: plugin/plugins/nopfs/nopfs.go
Comment (lidel) [inline plugin/plugins/nopfs/nopfs.go:1]: @hsanjuan quick  feedback: this PR needs a add a basic regression test  that confirms default support is wired ok in kubo. maybe create a denylist with one entry, start daemon, and
- confirm blocking of a single CID works (`ipfs block get` fails)
- confirm blocking is disabled when env variable is set (`block get` succeeds)


(adding test to https://github.com/ipfs/kubo/tree/master/test/cli should be bit easier these days, we don't use sharness for new things anymore)

Comment (hacdias) [reply plugin/plugins/nopfs/nopfs.go:1]: Yes, feel free to reach out if you have any questions 😄

**Comment by lidel** [comment]
Added docs, now I am testing + applying changes caused by recent boxo reactor, will open relevant PRs later today.

**Comment by lidel** [comment]
I've added tests `/ipfs/cid/*` rule: it works as expected on CLI, but fails on gateway.

Repro: blocking the root CID of wikipedia, and everything under it: 

```console
$ cat  $IPFS_PATH/denylists/text.deny
/ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze/*
```

It correctly blocks things in CLI:

```console
$ ipfs resolve /ipns/en.wikipedia-on-ipfs.org/wiki/
Error: /ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze/wiki is blocked and cannot be provided
````

but the gateway returns blocked payload, bypasses the rule somehow:

```console
$ curl -is http://127.0.0.1:8080/ipns/en.wikipedia-on-ipfs.org/wiki/ | grep HTTP
HTTP/1.1 200 OK
```

**Comment by hsanjuan** [comment]
> but the gateway returns blocked payload, bypasses the rule somehow:
> 
> ```
> $ curl -is http://127.0.0.1:8080/ipns/en.wikipedia-on-ipfs.org/wiki/ | grep HTTP
> HTTP/1.1 200 OK
> ```

Yes, I think this is because the gateway here is created without passing the Kubo resolver, but it creates one internally. This is why I added the WithResolver() option. On rainbow currently this works as expected:

```
curl http://127.0.0.1:8090/ipns/en.wikipedia-on-ipfs.org/wiki/
failed to resolve /ipns/en.wikipedia-on-ipfs.org/wiki/: /ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze is blocked and cannot be provided
```

We can fix this once latest boxo has bubbled to Kubo.

**Comment by hacdias** [comment]
@hsanjuan you can use Boxo's latest commit in main here. We allow Kubo to use a commit version of Boxo and then at release time we ensure that it has a tagged version.

**Comment by hsanjuan** [comment]
> @hsanjuan you can use Boxo's latest commit in main here. We allow Kubo to use a commit version of Boxo and then at release time we ensure that it has a tagged version.

ah ok, I thought path changes etc had not been bubbled. Can do.

File: core/corehttp/gateway.go [hunk 1]
```diff
@@ -132,7 +132,11 @@ func newGatewayBackend(n *core.IpfsNode) (gateway.IPFSBackend, error) {
 		}
 	}
 
-	backend, err := gateway.NewBlocksBackend(bserv, gateway.WithValueStore(vsRouting), gateway.WithNameSystem(nsys))
+	backend, err := gateway.NewBlocksBackend(bserv,
+		gateway.WithValueStore(vsRouting),
+		gateway.WithNameSystem(nsys),
+		gateway.WithResolver(n.UnixFSPathResolver),
```
Comment (lidel) [inline core/corehttp/gateway.go:138]: @hsanjuan  hm.. mind  we have offline exchange in two places ([here](https://github.com/ipfs/kubo/blob/89a476948b1d57cff3c7f07457aaccf2ea4feaf9/core/corehttp/gateway.go#L115) and [here](https://github.com/ipfs/kubo/blob/89a476948b1d57cff3c7f07457aaccf2ea4feaf9/core/corehttp/gateway.go#L82)). 

If I understand correctly, by using global path resolver here, that uses different (unlimited) exchange backend, we ignore the gateway-specific no-fetch preference and path resolution may cause raw blocks to be fetched from the network when it did not happen before 

If the content fits in a single final block, gateway will host third-party content  when user explicitly asked to not fetch third-party data.

Comment (hsanjuan) [reply core/corehttp/gateway.go:138]: in the libp2pGatewayOption I don't see an issue.

In the other one I see it... we should create the resolver using the fetcher that uses the offline blockservice. I'll make that change.

Comment (hsanjuan) [reply core/corehttp/gateway.go]: see b4b7ca71d

**Comment by hsanjuan** [comment]
@lidel I updated nopfs to a commit that has both https://github.com/ipfs-shipyard/nopfs/pull/28 and https://github.com/ipfs-shipyard/nopfs/pull/27 merged. Eventually we should produce tags for boxo, nopfs and bubble that.

**Comment by lidel** [comment]
Close, but not ready yet: content blocking is not applied to "no fetch" gateways.

https://github.com/ipfs/kubo/pull/10161/commits/c99068e3a7bf0f77d657fdc7897182d1aa796565 adds missing tests for "no fetch" gateways one can expose. In both cases the offline mode is done by passing custom  blockservice/exchange into path resolver, which means global path resolver that has nopfs intercept is not used,  and the content blocking does not happen on these gateways.

Needs to be fixed before merge, but at least now we have tests that fail  on this PR to surface the problem.

I'll look into this on Friday, but if I run out of time, this might slip from 0.24 to 0.25.

**Review by lidel — APPROVED** [review]
Switched this PR to commits that are in main branches of boxo and nopfs. 
TODOs are addressed, and tests pass. Gateway returns HTTP 410.

I am merging this to ensure it is included in [0.24-rc1](https://github.com/ipfs/kubo/issues/10043), allowing wider community to test and provide feedback. 
We will switch to tagged release versions before final 0.24.

Once again, thank you @hsanjuan for work on NOpfs and the spec, made this possible.
